packages/wonder-blocks-core/util/get-element-intersection.js

```diff
@@ -0,0 +1,130 @@
+ * Indicates the visibility of an item on the horizontal and vertical axes.
+ */
+export type Intersection = {|
+    /**
+     * The intersection on the horizontal axis.
+     */
+    horizontal: ?AxisIntersection,
+
+    /**
+     * The intersection on the vertical axis.
+     */
+    vertical: ?AxisIntersection,
+|};
+
+const UndeterminedIntersection: Intersection = {
+    horizontal: null,
+    vertical: null,
+};
+
+const FullIntersection: Intersection = {
+    horizontal: "within",
+    vertical: "within",
+};
+
+function getAxisIntersection(
+    intersectingRect: ClientRect | DOMRect,
+    boundsRect: ClientRect | DOMRect,
+    axis: "vertical" | "horizontal",
+): AxisIntersection {
+    const start = (rect) => (axis === "horizontal" ? rect.left : rect.top);
+    const end = (rect) => (axis === "horizontal" ? rect.right : rect.bottom);
+
+    if (end(intersectingRect) <= start(boundsRect)) {
+        return "before";
+    } else if (start(intersectingRect) >= end(boundsRect)) {
+        return "after";
+    } else {
+        return "within";
+    }
+}
+
+/**
+ * Determines whether of an element against a single other element.
+ * It is assumed that the element provided as `boundsElement` is an element that
+ * could obscure the given `element`.
+ */
+function getElementIntersectionAgainstParent(
+    intersectingRect: ClientRect | DOMRect,
+    boundsElement: Element,
+): Intersection {
+    const boundsRect = boundsElement.getBoundingClientRect();
+
+    const horizontal = getAxisIntersection(
+        intersectingRect,
+        boundsRect,
+        "horizontal",
+    );
+
+    const vertical = getAxisIntersection(
+        intersectingRect,
+        boundsRect,
+        "vertical",
+    );
+
+    return {horizontal, vertical};
+}
+
+/**
+ * Determines if a given element intersects with the visible bounds of its
+ * scroll parents, or the bounds of a specific element.
+ */
+export default function getElementIntersection(
+    element: Element,
+    boundsElement?: Element,
+): Intersection {
+    if (!element) {
+        // An non-existant element is definitely not visible.
+        return UndeterminedIntersection;
+    }
+
+    const intersectingRect = element.getBoundingClientRect();
+    // If we're looking against a single boundary element, then we just do that.
+    if (boundsElement) {
+        return getElementIntersectionAgainstParent(
+            intersectingRect,
+            boundsElement,
+        );
+    }
+
+    // Otherwise, we enumerate the scroll parents and test against those.
+    for (const scrollParent of enumerateScrollAncestors(element)) {
+        const intersection = getElementIntersectionAgainstParent(
+            intersectingRect,
+            scrollParent,
+        );
+        if (
+            intersection.vertical !== "within" ||
+            intersection.horizontal !== "within"
+        ) {
+            // Stop looking, we've found something that is hiding the element.
+            return intersection;
```

Comment:
Why not get rid fo the `if` statement and save some lines of code?

Reply:
Not sure I understand. We are looping through each parent and stopping on the first one that is hiding the element. Without the `if` we would just return on the first parent result, even if it was not hiding the element.

Perhaps I am misunderstanding your suggestion. Can you elaborate?

Reply:
I missed that this was inside a `for` loop and that it was causing the loop to terminate early.

Reply:
Yeah, I noticed it was a little...hard to read. Hopefully the updated commentary helps make it clearer?